source-code/sdk-js/src/helpers/ast.ts

```diff
@@ -0,0 +1,108 @@
+ * @param identifier The function identifier with which the declarator should be wrapped.
+ * @returns A promise that resolves to the resulting ast or rejects if the searched for declarator can't be found.
+ */
+export const wrapVariableDeclaration = (async (sourceAst, searchIdentifier, identifier) => {
+	const sourceAstClone = structuredClone(sourceAst)
+	let found = false
+	walk(sourceAstClone, {
+		enter(node) {
+			if (
+				node.type === "VariableDeclarator" &&
+				node.id.type === "Identifier" &&
+				node.id.name === searchIdentifier &&
+				node.init
+			) {
+				found = true
+				const newNode = structuredClone(node)
+				newNode.init = wrapWithCallExpression(node.init, identifier)
+				this.replace(newNode)
+				this.skip()
+			}
+		},
+	})
+	if (!found) throw new Error("Couldn't find variable declarator.")
+	return sourceAstClone
+}) satisfies wrapVariableDeclaration
+
+/**
+ * Inserts the provided ast at the given position in a source ast.
+ * Is immutable towards it's inputs, meaning it DOESN'T MANIPULATE THEM.
+ * @param sourceAst The ESTree Program to insert into.
+ * @param ast The ESTree Node to insert
+ * @param position The positional information for the insert.
+ * Should be an object containing a string array that is the path to the insertion point within the sourceAst.
+ * Could typically begin with ["body", ...] as that is one of the first keys within an ESTree Program
+ * For reference on how such a string could look, see https://lodash.com/docs/4.17.15#get
+ * @param position.after An array of strings denoting the path to the insertion point within the sourceAst, before which ast shall be inserted.
+ * Use array index 0 to insert into an empty Array.
+ * @param position.before An array of strings denoting the path to the insertion point within the sourceAst, after which ast shall be inserted.
+ * Use array index 0 to insert into an empty Array.
+ * @returns Promise that either resolves to the resulting ast or throws if the given insertion position couldn't be found
+ * or the positional information array length does not satisfy length % 3 === 2
+ */
+export const insertAst = (async (sourceAst, ast, { before, after }) => {
+	const sourceAstClone = structuredClone(sourceAst)
+	const position = after || before
+	if (position.length % 3 !== 2) {
+		throw new Error("The length of 'before' or 'after' has to be a multiple of two.")
+	}
+	const targetParent = get(sourceAstClone, position.slice(0, -2), sourceAstClone)
+	const targetArrayKey = position.at(-2)
+	const targetArrayIndex = Number(position.at(-1))
+	let newAst
+	walk(sourceAstClone, {
+		enter(node) {
+			const newNode = structuredClone(node)
+			if (node === targetParent) {
+				const targetArray = get(newNode, targetArrayKey as string, undefined) as Node[] | undefined
+				if (!targetArray) throw new Error("Couldn't access given position in AST.")
+				else targetArray.splice(targetArrayIndex + (before ? 0 : 1), 0, structuredClone(ast))
```

Comment:
should't this be `(position === before ? 0 : 1)` instead of `(before ? 0 : 1)`

Reply:
I am only checking if before is defined or not there because I want to add +1 if before is undefined. So I think it is still functionally correct. But if you say it's bad style, I'll be happy to change it

Reply:
This looks like a bug to me. You use `before` just to initialize `position`. But `position` won't be set to `before` if `after` is defined. And here you do some conditional logic if `before` got passed. Even if `position` is not set to `before`.

Reply:
ah, I haven't looked at the types ^^
Please change the code to
```diff
-const position = after || before
-const position = before || after
```
to be on the safe side